pvlib/mlfm.py

```diff
@@ -0,0 +1,804 @@
+G_LIC = 0.2   # LIC irradiance  [kW/m^2]
+
+
+#  Define standardised MLFM graph colours as a dict ``clr``
+
+clr = {
+    # parameter_clr colour            R   G   B
+    'irradiance':   'darkgreen',   # 000 064 000
+    'temp_module':  'red',         # 255 000 000
+    'temp_air':     'yellow',      # 245 245 220
+    'wind_speed':   'grey',        # 127 127 127
+
+    'i_sc':         'purple',      # 128 000 128
+    'r_sc':         'orange',      # 255 165 000
+    'i_ff':         'lightgreen',  # 144 238 144
+    'i_mp':         'green',       # 000 255 000
+    'i_v':          'black',       # 000 000 000 between i and v losses
+    'v_ff':         'cyan',        # 000 255 255
+    'v_mp':         'blue',        # 000 000 255
+    'r_oc':         'pink',        # 255 192 203
+    'v_oc':         'sienna',      # 160 082 045
+
+    'pr_dc':        'black',       # 000 000 000
+}
+
+
+def mlfm_meas_to_norm(dmeas, ref):
+    '''
+    Convert measured power, current and voltage to normalized values.
+
+    Parameters
+    ----------
+    dmeas : DataFrame
+        Measurements. Must include columns:
+
+        * `'poa_global'` global plane of array irradiance [W/m^2]
+        * `'temp_module'` module temperature [C]
+        * `'p_mp'` - power at maximum power point [W]
+
+        May include optional columns:
+
+        * `'i_sc'` - current at short circuit condition [A]
+        * `'v_oc'` - voltage at open circuit condition [V]
+        * `'i_mp'` - current at maximum power point [A]. Must be accompanied
+          by `'i_sc'`.
+        * `'v_mp'` - voltage at maximum power point [V]. Must be accompanied
+          by `'v_oc'`.
+        * `'r_sc'` - inverse of slope of IV curve at short circuit condition.
+          Requires both `'i_sc'` and `'v_oc'`. [Ohm]
+        * `'r_oc'` - inverse slope of IV curve at open circuit condition.
+          Requires both `'i_sc'` and `'v_oc'` [Ohm]
+
+    ref : dict
+        Reference values. Must include:
+
+        * `'p_mp'` - Power at maximum power point at Standard Test Condition
+          (STC). [W]
+        * `'gamma_pdc'` - Temperature coefficient of power at STC. [1/C]
+
+        May include:
+
+        * `'i_sc'` - Current at short circuit at STC. Required if `'i_sc'` is
+          present in ``dmeas``. [A]
+        * `'v_oc'` - Voltage at open circuit at STC. Required if `'v_oc'` is
+          present in ``dmeas``. [A]
+        * `'beta_v_oc'` - Temperature coefficient of open circuit voltage at
+          STC. Required if `'v_oc'` is present in ``dmeas``. [1/C]
+
+    Returns
+    -------
+    dnorm : DataFrame
+        Normalised values.
+
+        * `'pr_dc'` is `'p_mp'` normalised (divided) by reference `'p_mp'` \
+            and by `'poa_global'` in kW/m^2.
+        * `'pr_dc_temp_corr'` is `'pr_dc'` adjusted to 25C.
+        * Columns `'i_sc'`, `'i_mp'`, `'v_oc'`, `'v_mp'`, `'v_oc_temp_corr'`,
+          `'r_sc'`, `'r_oc'`, `'i_ff'`, `'v_ff'` are returned when the
```

Comment:
The naming convention here strikes me as inconsistent. These are also all normalized values for the DC circuit, but they don't follow a naming convention like `pr_dc` and `pr_dc_temp_corr`.

Reply:
pr is usually defined as the performance ratio ac as PR = Yf/Yr. 
There isn't a single naming convention for the equivalent at dc conditions Ya/Yr. 
I've seen both performance_factor, module_factor (which I don't like because it can be used for modules, strings, dc_combiners and inverter_inputs) so had used pr_ac and pr_dc to differentiate the two.

Reply:
Thanks for explaining the Performance Ratio connection that I totally missed (doh!). This might be worth a comment, unless I missed it.

My pattern-matching brain would prefer to just use `p_mp` and `p_mp_temp_corr` because (1) this is consistent with the remaining `i_sc`, `i_mp`, etc., and (2) the indication of "normalization" is already in the dataframe name `dnorm`.

Alternatively, one could apply some sort of consistent normalization pre/suffix to each variable in the dataframe.

Reply:
Thanks Mark. 
The problem I see is that on the dc side there are values for pr and p_mp (W or normalised), on the ac side there are also pr and p_mp (W or normalised) and these will differ by the factor inv_eff.
That's why I've used _dc and _ac suffixes to differentiate between values that would have the same name.

I've also used meas_ or norm_ as prefixes or dataframe names.

We could use p_dc and p_ac  but unless we have a dc equivalent of pr (ac) then I think I would need a pr_dc suffix.

I always use p_mp = i_mp * v_mp as at perfect mppt and p_dc = i_dc * v_dc or p_ac = i_ac * v_ac to cope with imperfect mpp tracking or clipping.